mpcs/src/util/merkle_tree.rs

```diff
@@ -23,36 +23,89 @@ use ark_std::{end_timer, start_timer};
 use super::hash::write_digest_to_transcript;
 
 #[derive(Clone, Debug, Default, Serialize, Deserialize)]
-#[serde(bound(serialize = "E: Serialize", deserialize = "E: DeserializeOwned"))]
-pub struct MerkleTree<E: ExtensionField>
+#[serde(bound(deserialize = "E: DeserializeOwned"))]
+pub struct MerkleTreeDigests<E: ExtensionField>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
     inner: Vec<Vec<Digest<E::BaseField>>>,
-    leaves: Vec<FieldType<E>>,
 }
 
-impl<E: ExtensionField> MerkleTree<E>
+impl<E: ExtensionField> MerkleTreeDigests<E>
 where
     E::BaseField: Serialize + DeserializeOwned,
 {
-    pub fn compute_inner(leaves: &FieldType<E>) -> Vec<Vec<Digest<E::BaseField>>> {
-        merkelize::<E>(&[leaves])
+    pub fn from_leaves(leaves: &FieldType<E>) -> Self {
+        Self {
+            inner: merkelize::<E>(&[leaves]),
+        }
+    }
+    pub fn from_leaves_base(leaves: &[E::BaseField]) -> Self {
+        Self {
+            inner: merkelize_base::<E>(&[leaves]),
+        }
+    }
+    pub fn from_leaves_ext(leaves: &[E]) -> Self {
+        Self {
+            inner: merkelize_ext::<E>(&[leaves]),
+        }
     }
 
-    pub fn compute_inner_base(leaves: &[E::BaseField]) -> Vec<Vec<Digest<E::BaseField>>> {
-        merkelize_base::<E>(&[leaves])
+    pub fn from_batch_leaves(leaves: &[&FieldType<E>]) -> Self {
+        Self {
+            inner: merkelize::<E>(leaves),
+        }
     }
 
-    pub fn compute_inner_ext(leaves: &[E]) -> Vec<Vec<Digest<E::BaseField>>> {
-        merkelize_ext::<E>(&[leaves])
+    pub fn root(&self) -> Digest<E::BaseField> {
+        self.inner.last().unwrap()[0].clone()
     }
 
-    pub fn root_from_inner(inner: &[Vec<Digest<E::BaseField>>]) -> Digest<E::BaseField> {
-        inner.last().unwrap()[0].clone()
+    pub fn root_ref(&self) -> &Digest<E::BaseField> {
+        &self.inner.last().unwrap()[0]
```

Comment:
It looks like we require that `inner` not be empty.  I suggest [we encode that in the type](https://docs.rs/nonempty/latest/nonempty/), instead of going with `Vec` which is happy being empty.

Reply:
Yes, that would be better. But there are more to constrain than this, e.g., the size of the vectors should be 1, 2, 4, 8, and so on. I don't know how to encode all these in the type. Currently, it's just relying on `inner` being private, and is never changed once the Merkle tree is built.